src/java.base/share/classes/java/text/ListFormat.java

```diff
@@ -0,0 +1,593 @@
+    private static final Pattern PARSE_MIDDLE = Pattern.compile("\\{0}(.*?)\\{1}");
+    private static final Pattern PARSE_END = Pattern.compile("\\{0}(.*?)\\{1}(.*?)");
+    private static final Pattern PARSE_TWO = Pattern.compile("(.*?)\\{0}(.*?)\\{1}(.*?)");
+    private static final Pattern PARSE_THREE = Pattern.compile("(.*?)\\{0}(.*?)\\{1}(.*?)\\{2}(.*?)");
+    private transient Pattern startPattern;
+    private transient String middleBetween;
+    private transient Pattern endPattern;
+
+    private ListFormat(Locale l, String[] patterns) {
+        locale = l;
+        this.patterns = patterns;
+        init();
+    }
+
+    private void init() {
+        // get pattern strings
+        var m = PARSE_START.matcher(patterns[START]);
+        String startBefore;
+        String startBetween;
+        if (m.matches()) {
+            startBefore = m.group(1);
+            startBetween = m.group(2);
+        } else {
+            throw new IllegalArgumentException("start pattern is incorrect: " + patterns[START]);
+        }
+        m = PARSE_MIDDLE.matcher(patterns[MIDDLE]);
+        if (m.matches()) {
+            middleBetween = m.group(1);
+        } else {
+            throw new IllegalArgumentException("middle pattern is incorrect: " + patterns[MIDDLE]);
+        }
+        m = PARSE_END.matcher(patterns[END]);
+        String endBetween;
+        String endAfter;
+        if (m.matches()) {
+            endBetween = m.group(1);
+            endAfter = m.group(2);
+        } else {
+            throw new IllegalArgumentException("end pattern is incorrect: " + patterns[END]);
+        }
+
+        // Validate two/three patterns, if given. Otherwise, generate them
+        if (patterns[TWO] != null && !patterns[TWO].isEmpty()) {
+            if (!PARSE_TWO.matcher(patterns[TWO]).matches()) {
+                throw new IllegalArgumentException("pattern for two is incorrect: " + patterns[TWO]);
+            }
+        } else {
+            patterns[TWO] = startBefore + "{0}" + endBetween + "{1}" + endAfter;
+        }
+        if (patterns[THREE] != null && !patterns[THREE].isEmpty()) {
+            if (!PARSE_THREE.matcher(patterns[THREE]).matches()) {
+                throw new IllegalArgumentException("pattern for three is incorrect: " + patterns[THREE]);
+            }
+        } else {
+            patterns[THREE] = startBefore + "{0}" + startBetween + "{1}" + endBetween + "{2}" + endAfter;
+        }
+
+        startPattern = Pattern.compile(startBefore + "(.+?)" + startBetween);
+        endPattern = Pattern.compile(endBetween + "(.+?)" + endAfter);
+    }
+
+    /**
+     * {@return the available locales that support list formatting}
+     */
+    public static Locale[] getAvailableLocales() {
+        // Same as a typical format class
+        return DateFormat.getAvailableLocales();
+    }
+
+    /**
+     * {@return the list format object for the default
```

Comment:
"list format" will read better using the class name `ListFormat`. Here an in other `getInstance` methods.
As is, in prose it looks like it returns some kind of list.